### PR TITLE
Fix: NUMA comm split type 

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**
With the newer version of hwloc, `get_ancestor_by_type` doesn't list NUMA objects when queried with `HWLOC_OBJ_NUMANODE` type objects because NUMA objects are not assigned depths in topology tree. We need to query the NUMA objects and check if the cpuset of a process is covered by any of the NUMA objects.

### CODE REVIEW

File: src/mpi/comm/comm_split_type.c [hunk 1]
```diff
@@ -173,8 +173,22 @@ static int node_split_processor(MPIR_Comm * comm_ptr, int key, const char *hintv
...
         color = obj_containing_cpuset->logical_index;
     } else {
-        hwloc_obj_t hobj = hwloc_get_ancestor_obj_by_type(MPIR_Process.topology, query_obj_type,
-                                                          obj_containing_cpuset);
+        hwloc_obj_t hobj = NULL;
+        if (query_obj_type == HWLOC_OBJ_NUMANODE) {
+            hwloc_obj_t tmp = NULL;
+            while ((tmp =
+                    hwloc_get_next_obj_by_type(MPIR_Process.topology, query_obj_type,
+                                               tmp)) != NULL) {
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: I'm not sure I follow the logic here.  When you pass `NULL` to `hwloc_get_next_obj_by_type` in the third parameter, what's the expected return value?

File: src/mpi/comm/comm_split_type.c [hunk 2]
```diff
@@ -173,8 +173,22 @@ static int node_split_processor(MPIR_Comm * comm_ptr, int key, const char *hintv
...
+                if (hwloc_bitmap_isincluded(obj_containing_cpuset->cpuset, tmp->cpuset) ||
+                    hwloc_bitmap_isequal(tmp->cpuset, obj_containing_cpuset->cpuset)) {
+                    hobj = tmp;
+                    break;
+                }
+            }
+        } else {
+            hobj = hwloc_get_ancestor_obj_by_type(MPIR_Process.topology, query_obj_type,
+                                                  obj_containing_cpuset);
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: Why is this needed?  Why can't the above logic in the `if` branch work also for non-NUMANODE objects too?

**Comment by pavanbalaji** [comment]
Wasn't this working in `hwloc-2.0`?  Which version of `hwloc` did this change in?  It seems very surprising to me that they broke it since NUMA nodes are very common.

File: src/mpi/comm/comm_split_type.c (hunk 1 above)
Comment (KavithaTipturMadhu) [reply src/mpi/comm/comm_split_type.c]: hwloc_get_next_obj_by_type is like an iterator that takes NULL initially and returns the first object of the type `query_obj_type`. We need to pass the previous object to move ahead, subsequently.

File: src/mpi/comm/comm_split_type.c (hunk 2 above)
Comment (KavithaTipturMadhu) [reply src/mpi/comm/comm_split_type.c]: The same logic above works, I'll remove the else block.

**Comment by KavithaTipturMadhu** [comment]
This wasn't working with NUMA nodes with hwloc-2.0 update, I seem to have missed testing this. The latest release conforms to [this](https://www.open-mpi.org/projects/hwloc/doc/v2.0.1/a00327.php#upgrade_to_api_2x_memory_find) organization of NUMA nodes.

File: src/mpi/comm/comm_split_type.c (hunk 2 above)
Comment (KavithaTipturMadhu) [reply src/mpi/comm/comm_split_type.c]: I have updated the patch with the else block removed and tested the changes too.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
One of tcp tests is failing because of freebsd communication connection failure. I checked ucx test manually and it seems to be running fine, hence rerunning ucx tests.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch4/ucx

**Comment by KavithaTipturMadhu** [comment]
All tests are green other than one because of freebsd connection failure. @pavanbalaji Shall I merge this?

**Comment by pavanbalaji** [comment]
Can you add a comment saying that you cannot use `hwloc_get_ancestor_obj_by_type` because of the NUMA node issue?  Point to the web link on the `hwloc` website mentioning this.  No need to run the tests again once you add the comment.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most
